search/search_repository_whitebox_test.go

```diff
@@ -45,189 +53,210 @@ func (s *searchRepositoryWhiteboxTest) SetupTest() {
 }
 
 type SearchTestDescriptor struct {
-	wi             workitem.WorkItem
+	fields         map[string]interface{}
 	searchString   string
 	minimumResults int
 }
 
-func (s *searchRepositoryWhiteboxTest) TestSearchByText() {
-	wir := workitem.NewWorkItemRepository(s.DB)
-
+func (s *searchRepositoryWhiteboxTest) setupTestDataSet() ([]SearchTestDescriptor, *tf.TestFixture) {
+	// given
 	testDataSet := []SearchTestDescriptor{
 		{
-			wi: workitem.WorkItem{
-				Fields: map[string]interface{}{
-					workitem.SystemTitle:       "test sbose title '12345678asdfgh'",
-					workitem.SystemDescription: rendering.NewMarkupContentFromLegacy(`"description" for search test`),
-					workitem.SystemCreator:     "sbose78",
-					workitem.SystemAssignees:   []string{"pranav"},
-					workitem.SystemState:       "closed",
-				},
+			fields: map[string]interface{}{
+				workitem.SystemTitle:       "test sbose title '12345678asdfgh'",
+				workitem.SystemDescription: rendering.NewMarkupContentFromLegacy(`"description" for search test`),
+				workitem.SystemCreator:     "sbose78",
```

Comment:
At first I wanted to suggest to setup all workitems like this:

```go
fxt := tf.NewTestFixture(s.T(), s.SB,
  tf.Identities(3, tf.SetIdentityUsernames([]string{"sbose78", "pranav", "pgore"})),
  tf.WorkItems(7, func(fxt *tf.TestFixture, idx int) error{
    wi := fxt.WorkItems[idx]
    // almost all WIs are created by sbose, except a few
    wi.Fields[workitem.SystemCreator] = fxt.IdentitiesByUsername("sbose78")
    // almost all WIs are closed (exceptions are specified below)
    wi.Fields[workitem.SystemState} = workitem.SystemStateClosed
    // all WIs are assigned to pranav
    wi.Fields[workitem.SystemAssignees] = []string{fxt.IdentitiesByUsername("pranav").String()}
  
    switch idx {
    case 0:
      wi.Fields[workitem.SystemTitle] = "test sbose title '12345678asdfgh'"
      wi.Fields[workitem.SystemDescription] = rendering.NewMarkupContentFromLegacy(`"description" for search test`)
    case 1:
      wi.Fields[workitem.SystemTitle] = "add new error types in models/errors.go'"
      wi.Fields[workitem.SystemDescription] = rendering.NewMarkupContentFromLegacy(`Make sure remoteworkitem can access..`)
    case 2: 
      wi.Fields[workitem.SystemTitle] = "test sbose title '12345678asdfgh'"
      wi.Fields[workitem.SystemDescription] = rendering.NewMarkupContentFromLegacy(`"description" for search test`)
    case 3:
      // will test behaviour when null fields are present. In this case, "system.description" is nil
      wi.Fields[workitem.SystemTitle] = "test nofield sbose title '12345678asdfgh'"
      // no description
    case 4:
      // will test behaviour when null fields are present. In this case, "system.description" is nil
      wi.Fields[workitem.SystemTitle] = "test should return 0 results'"
      // no description
    case 5:
      // search stirng with braces should be acceptable case
      wi.Fields[workitem.SystemTitle] = "Bug reported by administrator for input = (value)"
      // no description
      wi.Fields[workitem.SystemCreator] = fxt.IdentitiesByUsername("pranav")
      wi.Fields[workitem.SystemState} = workitem.SystemStateNew
    case 6:
      // search stirng with surrounding braces should be acceptable case
      wi.Fields[workitem.SystemTitle] = "trial for braces (pranav) {shoubhik} [aslak]"
      // no description
      wi.Fields[workitem.SystemCreator] = fxt.IdentitiesByUsername("pgore")
      wi.Fields[workitem.SystemState} = workitem.SystemStateNew
    }
    return nil
  }),
)
```

But then I realized that it is probably much easier to just create a fixture for each individual test that is formulated by the test data. Alongside the `fields`, `searchString`, `minimumResults` you could add a slice of recipe functions. Then it is much easier to see what each test searches for, don't you think? 

Sorry, but I'm not sure if I did express my ideas here properly ;) They essentially evolve around: It is good to have table driven tests, but only if the table entry is easy to read and the scenario is easily understood and don't interfere with each other. Here we create a very complex scenario (aka creation of work items), and each time we add a search to that table, we influence the rest because the work items are all created in one space. It might be that at one point we have WIs `A`, `B`, and `C` and a search test requires us to find one work item (let's say `A`), so `minimumResuls` equals `1`. Then we modify the search functionality and add another WI: `D`. By our modification we no longer find `A` but `D`. That said, our first test is now broken. but it doesn't report to be broken because we still find a new work item: `D`.

I hope it is clear now why we need to have an individual test fixture for each `searchString` that we test. Then we can be sure that each new `searchString` that we test that would normally add new entry in the `testData` table, now doesn't modify the old search tests.

Reply:
My latest commit proposes a refactoring of the `Search` test by using sub-tests. Please let me know what you think about it.